Fix hooks handling for unpickled nnmodule (try 2)

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96311
* __->__ #96310
* #96238

Ensure _update_has_hooks() will work regardless of which
hook dicts exist on the module.

Second try after revert of https://github.com/pytorch/pytorch/pull/96224 where unpickled module didn't have _backward_hooks field.

wconstab committed:
Fix hooks handling for unpickled nnmodule (try 2)

Ensure _update_has_hooks() will work regardless of which
hook dicts exist on the module.

[ghstack-poisoned]

ezyang commented:
Uh, seems unlikely. But alban put a module BC test PR https://github.com/pytorch/pytorch/pull/96238 which you should rebase after prolly

wconstab committed:
Update on "Fix hooks handling for unpickled nnmodule (try 2)"


Ensure _update_has_hooks() will work regardless of which
hook dicts exist on the module.

Second try after revert of https://github.com/pytorch/pytorch/pull/96224 where unpickled module didn't have _backward_hooks field.

[ghstack-poisoned]
